elements/lisk-transactions/src/sign.ts

```diff
@@ -243,63 +208,62 @@ export const signMultiSignatureTransactionWithPrivateKey = (
 	if (validationErrors) {
 		throw validationErrors;
 	}
-	// Sort keys
+
 	keys.mandatoryKeys.sort((publicKeyA, publicKeyB) => publicKeyA.compare(publicKeyB));
 	keys.optionalKeys.sort((publicKeyA, publicKeyB) => publicKeyA.compare(publicKeyB));
 
-	const publicKey = ed.getPublicKeyFromPrivateKey(privateKey);
 	const signature = ed.signDataWithPrivateKey(
 		TAG_TRANSACTION,
 		chainID,
 		getSigningBytes(transactionObject, paramsSchema),
 		privateKey,
 	);
 
-	if (
-		includeSenderSignature &&
-		Buffer.isBuffer(transactionObject.senderPublicKey) &&
-		publicKey.equals(transactionObject.senderPublicKey)
-	) {
-		// eslint-disable-next-line no-param-reassign
-		transactionObject.signatures[0] = signature;
-	}
-
-	// Locate where this public key should go in the signatures array
-	const mandatoryKeyIndex = keys.mandatoryKeys.findIndex(aPublicKey =>
-		aPublicKey.equals(publicKey),
-	);
-	const optionalKeyIndex = keys.optionalKeys.findIndex(aPublicKey => aPublicKey.equals(publicKey));
-
-	// If it's a mandatory Public Key find where to add the signature
-	if (mandatoryKeyIndex !== -1) {
-		const signatureOffset = includeSenderSignature ? 1 : 0;
-		// eslint-disable-next-line no-param-reassign
-		transactionObject.signatures[mandatoryKeyIndex + signatureOffset] = signature;
-	}
+	const publicKey = ed.getPublicKeyFromPrivateKey(privateKey);
+	const accountKeys = keys.mandatoryKeys.concat(keys.optionalKeys);
 
-	if (optionalKeyIndex !== -1) {
-		const signatureOffset = includeSenderSignature ? 1 : 0;
-		// eslint-disable-next-line no-param-reassign
-		transactionObject.signatures[keys.mandatoryKeys.length + optionalKeyIndex + signatureOffset] =
-			signature;
+	// Find the position for the signature in the signatures array, based on the public key of the signer
+	for (let i = 0; i < accountKeys.length; i += 1) {
+		if (accountKeys[i].equals(publicKey)) {
+			// eslint-disable-next-line no-param-reassign
+			transactionObject.signatures[i] = signature;
+		} else if (transactionObject.signatures[i] === undefined) {
+			// eslint-disable-next-line no-param-reassign
+			transactionObject.signatures[i] = Buffer.alloc(0);
+		}
 	}
 
-	sanitizeSignaturesArray(transactionObject, keys, includeSenderSignature);
-
 	return { ...transactionObject, id: utils.hash(getBytes(transactionObject, paramsSchema)) };
 };
 
+// TODO: Check if these backward compatibility aliases are still needed
```

Comment:
I think it's better to create an issue when we create TODO

Reply:
Good point 👍🏻 

Created issue #7992

Reply:
did this come back again? 🤔

Reply:
Ahhh, force pushed and didn't change that 😅

I'll force push again...